### DISCUSSION TITLE
[GSoC] Julia wrapper

### DISCUSSION THREAD

**Discussion by johnnychen94** [halide/Halide#5824] [General]
Hi,

My name is Jiuning Chen(Johnny Chen), I'm a second-year Ph.D. student in ECNU(East China Normal University), a GSoC 19 student in Julia, and also a core maintainer of [JuliaImages](https://github.com/JuliaImages/Images.jl) since that. My previous main research interest is low-level image processing (e.g., denoise) using filter techs.

Both Julia and halide have shown their great advantages in image processing and I want to see if I can try to get the best of the two worlds by providing a Julia interface to Halide. 

I've learned a lot from the halide paper when I try to optimize my Julia codes but never tried halide; it was quite a long time ago since my last usage of C++. To apply for the [Halide Code Generation For More Languages](https://github.com/halide/Halide/wiki/%5BGSoC-2021%5D-Halide-Code-Generation-For-More-Languages) GSoC project, I definitely need some refresh of my C++ knowledge, I'm wondering if you could provide some easy-to-start issues that I can work on to get a better understanding of halide (and of course C++) before I start to write my proposal.

Also, is this issue an appropriate place to update progress, or should I contact you via other channels/apps?

Best regards

cc: @zvookin @rootjalex @BachiLi

**Comment by rootjalex** [comment]
If you haven't already tried these out, the[ tutorials](https://halide-lang.org/tutorials/tutorial_introduction.html) are a good way to get more familiar with Halide.

I'm not sure about specific issues that are relevant to you at the moment, but Tzu-Mao recommended trying to understand our [PyTorch codegen](https://github.com/halide/Halide/blob/master/src/CodeGen_PyTorch.cpp) [module](https://github.com/halide/Halide/blob/master/src/CodeGen_PyTorch.h) as a good starting point.

The [gitter](https://gitter.im/halide/Halide) is a good communication method that we use.

**Comment by rootjalex** [comment]
A bit more on this, specific to Julia: after a quick glance over [Cxx.jl](https://github.com/JuliaInterop/Cxx.jl) and [CxxWrap.jl](https://github.com/JuliaInterop/CxxWrap.jl), CxxWrap seems to be the more plausible option for embedding (if there are others who are more familiar with Julia, please correct me). [This](https://github.com/JuliaInterop/CxxWrap.jl/issues/127) issue links to some examples that may be relevant

**Comment by johnnychen94** [comment]
I plan to follow a similar route [OpenCV-Julia binding project](https://github.com/opencv/opencv/wiki/OE-31.-Julia-Bindings) did since it's written in C++, too. So yeah I'll use CxxWrap. But before that, I'll need to try and learn the halide internals to understand it. I'll do a halide-Julia benchmark first in the next few days. Hopefully, I can put my hands on the wrapper project next weekend.

**Reply by johnnychen94** [reply]
I started by trying to integrate it in Linux first.

[CxxWrap_example](https://github.com/johnnychen94/CxxWrap_example) is my first attempt to test if I'm making things correctly, and it works. So here's my plan for the halide-julia project. To distinguish the halide itself and the Julia binding part, I'll call the Halide main library `libhalide` and the Julia part `libhalide-julia`.

1. Unlike the pytorch binding, I plan to separate the Julia binding from Halide's main repo, so I'll make a `libhalide-julia` thing, this is the C++ part of the CxxWrap.
2. Then a `Halide.jl` package that provides the interface to Julia users and passes Halide calls from `libhalide-julia` to `libhalide`.
3. After validating step 2, set up the [BinaryBuilder](https://github.com/JuliaPackaging/BinaryBuilder.jl) library building workflow and ship artifacts `libhalide` and `libhalide-julia` in [Yggdrasil](https://github.com/JuliaPackaging/Yggdrasil), which allows user to do `pkg> add Halide` in Julia and automatically installs the built artifacts.

Now there comes to several built issues when I scratch my `CMakeLists.txt` for `libhalide-julia`. My limited understanding of Halide, LLVM and CMake doesn't give a clear answer so I wish I can get some help here:

_Julia itself uses its [own flavor of LLVM](https://github.com/JuliaPackaging/Yggdrasil/blob/master/L/LLVM/common.jl#L236-L241)_ and it seems to be problematic for Halide (See the [TVM issue](https://github.com/JuliaLang/julia/issues/36342)). Thus I've started to build Halide with the Julia shared LLVM library, more specifically, I'm using `LLVM_full.v11.0.1.x86_64-linux-gnu-cxx11.tar.gz` from [LLVM_full_jll releases](https://github.com/JuliaBinaryWrappers/LLVM_full_jll.jl/releases).

Here's my manual build script:

```bash
export JULIA_DEPOT_PATH=$PWD/build/julia-depot
export LLVM_ROOT=$JULIA_DEPOT_PATH/artifacts/a8525442937e3f1f91ae295e6e001495ce7b5c18/
export LLVM_CONFIG=$LLVM_ROOT/bin/llvm-config

julia --project=libhalide-julia/ -e 'using Pkg; pkg"add LLVM_full_jll@11.0.1"'
cp -r $LLVM_ROOT/tools/* $LLVM_ROOT/bin

# git clone https://github.com/halide/Halide.git halide
cmake -DCMAKE_BUILD_TYPE=Release \
    -DLLVM_DIR="$LLVM_ROOT/lib/cmake/llvm" \
    -DHalide_SHARED_LLVM="ON" \
    _DTARGET_X86="ON" \
    -DTARGET_WEBASSEMBLY="OFF" \
    -DTARGET_AARCH64="OFF" \
    -DTARGET_ARM="OFF" \
    -DTARGET_AMDGPU="OFF" \
    -DTARGET_HEXAGON="OFF" \
    -DTARGET_MIPS="OFF" \
    -DTARGET_NVPTX="OFF" \
    -DTARGET_POWERPC="OFF" \
    -DTARGET_RISCV="OFF" \
    -DWITH_PYTHON_BINDINGS="OFF" \
    -DTARGET_D3D12COMPUTE="OFF" \
    -DTARGET_OPENCL="OFF" \
    -DTARGET_METAL="OFF" \
    -S halide -B build

cmake --build build -j48
```

I deliberately turned off every backend except for `X86` but it still says that other backends are enabled. Is this because I'm using a shared LLVM library?

```text
-- Found LLVM 11.0.1jl
-- Using LLVMConfig.cmake in: /mnt/cache/home/jc/Projects/Halide.jl/build/julia-depot/artifacts/a8525442937e3f1f91ae295e6e001495ce7b5c18//lib/cmake/llvm
-- Enabling AArch64 backend
-- Enabling AMDGPU backend
-- Enabling ARM backend
-- Enabling Hexagon backend
-- Enabling Mips backend
-- Enabling NVPTX backend
-- Enabling PowerPC backend
-- Enabling RISCV backend
-- Enabling WebAssembly backend
-- Enabling X86 backend
-- Compiling Halide WITHOUT RTTI.
-- Compiling Halide WITH exceptions.
-- Building autoschedulers enabled
-- Building tests enabled
-- Building Python bindings disabled
-- Building tutorials enabled
-- Building docs disabled
-- Building utils enabled
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/cache/home/jc/Projects/Halide.jl/build
```

Also, the Halide specific backends OpenCL, Meta and Direct3D 12 Compute are disabled, otherwise, it will throw undefined symbol error during linking.

Now the good news I get so far is that with this specific build setup, it passes all tests except one cross-compilation test, will this be an issue?

```text
99% tests passed, 1 tests failed out of 478
...
The following tests FAILED:
         81 - correctness_cross_compilation (Subprocess aborted)
```

FWIW, BinaryBuilder does a set of cross-complation in JuliaLang's own Linux bots; only MinGW is available for Windows targets. According to the Halide README.md, it seems that building halide on Windows only supports visual studio 2019, does this mean that building windows artifacts using binarybuilder is not possible?

**Reply by rootjalex** [reply]
I can't answer some of your build questions (I don't use CMake myself, but probably should make the switch from using make at some point...). Perhaps @alexreinking  @steven-johnson will be able to help there?

As for the project proposal: Your Greet.jl example looks like good for getting more familiar with CxxWrap! 
1. I worry about having the project repo as separate from the main repo, mostly due to the need to update it's Halide sub-repository, integrate it's tests with some build system, etc. etc. It's not a huge issue, but I'm curious what the other mentors think.
2. and 3. look good to me.

As for your `correctness_cross_compilation` failure, I can speak to that - it's likely due to using LLVM 11. I think I had issues with that test on my personal laptop before upgrading to LLVM 12. I can't remember exactly what the issue was, but I believe it had to do with the WASM build or something like that. Is Julia only using LLVM 11?

re: Windows and BinaryBuilder, Once again, going to defer to @alexreinking ? (or others who use/know anything about the Windows build system)

This is fantastic progress!

**Reply by johnnychen94** [reply]
> Is Julia only using LLVM 11?

Yes, according to [version.make](https://github.com/JuliaLang/julia/blob/f25d9885f6c55f6efe75abca5a9577f6bef07bed/deps/Versions.make#L47-L50), Julia 1.6.0 (released two days ago) is using LLVM 11.0.1. This is why I choose the same jll package; to try to avoid unexpected issues before I set up the whole pass pipeline from Julia to Halide.

I could build up Julia using LLVM 12 if it's necessary([MLIR.jl](https://github.com/vchuravy/MLIR.jl) indicates that this is doable), but the whole Julia workflow ecosystem is built upon their "official" release and it is quite problematic to use an external LLVM dep. For example, pacman in Arch builds Julia itself using the distro LLVM but it is known to be buggy. Given that Halide is more flexible, I chose to build halide with the specific LLVM version that Julia uses.

> I worry about having the project repo as separate from the main repo, mostly due to the need to update it's Halide sub-repository, integrate it's tests with some build system, etc. etc. It's not a huge issue, but I'm curious what the other mentors think.

Two C++ libraries will be built here:

- `libhalide`: this is the main repo and can be built using existing cmake options.
- `libhalide-julia`: this is required to be a shared library. Building this requires Julia binary, Julia's LLVM, [libcxxwrap-julia](https://github.com/JuliaInterop/libcxxwrap-julia), and CxxWrap.

The reason I plan to put the project in a separate repo is that I'm worried that 1) requiring special LLVM versions breaks the compatibility of halide if put on main repo. 2) setting up the CI workflow for this requires extra deps(Julia, BinaryBuilder, and etc.) and thus vulnerable somehow. Thus the plan I have in mind is to separate the build of `libhalide`(using Julia's LLVM) and the build of `libhalide-julia`. The former can be quite complicated when upstream Julia's LLVM version changes.

Another personal reason is... I'm really not a C++ guy so it terrifies me to work on such a big project :(

Ideally, part of the CI things can be delegated to Julialang's own CI bots if the workflow goes to BinaryBuilder.

**Reply by alexreinking** [reply]
> only MinGW is available for Windows targets. According to the Halide README.md, it seems that building halide on Windows only supports visual studio 2019, does this mean that building windows artifacts using binarybuilder is not possible?

Correct, we dropped support for MinGW a while back.

**Reply by alexreinking** [reply]
> Now the good news I get so far is that with this specific build setup, it passes all tests except one cross-compilation test, will this be an issue?

Any failing test is an issue.

**Reply by alexreinking** [reply]
> CxxWrap_example is my first attempt to test if I'm making things correctly, and it works.

Btw - before contributing any CMake code to Halide, you should read our [README_cmake.md](https://github.com/halide/Halide/blob/master/README_cmake.md#contributing-cmake-code-to-halide). In particular note our minimum version requirement of 3.16 and absolute prohibition of directory commands like `include_directories`.

**Reply by alexreinking** [reply]
> I worry about having the project repo as separate from the main repo, mostly due to the need to update it's Halide sub-repository, integrate it's tests with some build system, etc. etc. It's not a huge issue, but I'm curious what the other mentors think.

For anything GSOC, the aim is to introduce first-party support. That means upstreaming and setting up CI on our systems. So a separate repo is fine for rapid development, but it must be integrated into our repository in the end. From a workflow perspective, I would suggest forking Halide and working on a branch in the fork.

**Reply by alexreinking** [reply]
> I deliberately turned off every backend except for `X86` but it still says that other backends are enabled. Is this because I'm using a shared LLVM library?

Yes. When linking to shared LLVM, we enable all the backends that are available.

> Also, the Halide specific backends OpenCL, Meta and Direct3D 12 Compute are disabled, otherwise, it will throw undefined symbol error during linking.

Can you open an issue about this? Also, can you open an issue about the failing test with the Julia LLVM? Please provide full reproduction steps.

**Comment by johnnychen94** [comment]
Sorry if I'm asking the wrong question. Does Halide provide a C API? If so it's very possible that we could get rid of CxxWrap dep and thus make things easier.

**Reply by alexreinking** [reply]
No it does not.

**Comment by johnnychen94** [comment]
I see both `python_bindings` and `CodeGen_Pytorch` in this repo, my understanding for them is: `CodeGen_Pytorch` is doable because PyTorch has a C++ API that halide can directly communicate with. And Python in general needs pybind to wrap and pass the C+ codes, which is why `python_bindings` exists.

It seems that if I take the CxxWrap way, I'm doing a CxxWrap-version of `python_bindings` and thus make it available to call libhalide inside Julia.

**Reply by johnnychen94** [reply]
> it must be integrated into our repository in the end.

Actually, [`c45be97`](https://github.com/johnnychen94/Halide/commit/c45be97ecc0638484d6ccb171fd915d44b10d710) is what I've got so far with the suggested folder structure and now I get one simple method passed:

```julia
$ julia --project=julia_bindings/Halide

julia> using Halide

julia> f = Halide.Func("Blur")
Halide.FuncAllocated(Ptr{Nothing} @0x00000000028ee750)

julia> Halide.name(f)[] # [] is to dereference the C++ string reference
"Blur"
```

To make a real working Halide demo (e.g., [minimal JIT compiled example](https://github.com/halide/Halide/wiki/Minimal-JIT-compiled-example)), there're more types that needed to be wrapped. My plan is to finish the concept validation with a working demo before April 1, and then come up with a detailed proposal and send it to the mentioned mail `halide-gsoc2021-mentors@mit.edu`

The folder structure of it is:

```
Halide/julia_bindings/
├── CMakeLists.txt
├── Halide
│   ├── Manifest.toml
│   ├── Project.toml
│   └── src
│       └── Halide.jl
└── src
    ├── CMakeLists.txt
    ├── JlFunc.cpp
    ├── JlFunc.h
    ├── JlHalide.cpp
    └── JlHalide.h
```

Building Halide with Julia's shared LLVM library is a little bit puzzling so I chose to build it with a self-build version of LLVM 11.0.1 and keep using it until I actually hit symbolic issues mentioned in [the TVM issue](https://github.com/JuliaLang/julia/issues/36342).

**Reply by rootjalex** [reply]
> > it must be integrated into our repository in the end.
> 
> Actually, [`c45be97`](https://github.com/johnnychen94/Halide/commit/c45be97ecc0638484d6ccb171fd915d44b10d710) is what I've got so far with the suggested folder structure and now I get one simple method passed:
> 
> ```julia
> $ julia --project=julia_bindings/Halide
> 
> julia> using Halide
> 
> julia> f = Halide.Func("Blur")
> Halide.FuncAllocated(Ptr{Nothing} @0x00000000028ee750)
> 
> julia> Halide.name(f)[] # [] is to dereference the C++ string reference
> "Blur"
> ```

Looks great!

> To make a real working Halide demo (e.g., [minimal JIT compiled example](https://github.com/halide/Halide/wiki/Minimal-JIT-compiled-example)), there're more types that needed to be wrapped. My plan is to finish the concept validation with a working demo before April 1, and then come up with a detailed proposal and send it to the mentioned mail `halide-gsoc2021-mentors@mit.edu`

👍 

> The folder structure of it is:
> 
> ```
> Halide/julia_bindings/
> ├── CMakeLists.txt
> ├── Halide
> │   ├── Manifest.toml
> │   ├── Project.toml
> │   └── src
> │       └── Halide.jl
> └── src
>     ├── CMakeLists.txt
>     ├── JlFunc.cpp
>     ├── JlFunc.h
>     ├── JlHalide.cpp
>     └── JlHalide.h
> ```

Also looks good - generally, I would advise having as-similar-as-possible structure to our [python bindings](https://github.com/halide/Halide/tree/master/python_bindings), but I won't pretend to know all of the intricacies of Julia's build system.

> Building Halide with Julia's shared LLVM library is a little bit puzzling so I chose to build it with a self-build version of LLVM 11.0.1 and keep using it until I actually hit symbolic issues mentioned in [the TVM issue](https://github.com/JuliaLang/julia/issues/36342).

Seems reasonable for now!

**Reply by rootjalex** [reply]
Oh, I missed the higher-level comment. Personally, I agree, I think Julia should be more `python_bindings` style (not sure if others agree/disagree), as opposed to `CodeGen_Pytorch`.

**Reply by alexreinking** [reply]
@rootjalex those are actually different use-cases and they're both important. `python_bindings` exists to allow writing Halide programs from Python. `CodeGen_PyTorch` exists to allow compiled Halide pipelines to be used as PyTorch steps, without requiring a Halide installation for the end user. 

When we were drafting the proposal for this GSOC project, the intent was that the latter was actually higher priority, but having both would be great.

**Reply by steven-johnson** [reply]
> Personally, I agree, I think Julia should be more python_bindings style (not sure if others agree/disagree), as opposed to CodeGen_Pytorch

Agree 100%

**Comment by johnnychen94** [comment]
> _Julia itself uses its [own flavor of LLVM](https://github.com/JuliaPackaging/Yggdrasil/blob/master/L/LLVM/common.jl#L236-L241)_ and it seems to be problematic for Halide (See the [TVM issue](https://github.com/JuliaLang/julia/issues/36342)).

@vchuravy If it's not too complicated, may I ask will this be an issue for Halide? Or I have to build Julia from scratch just like https://github.com/vchuravy/MLIR.jl? I'm not familiar with LLVM at all so have no idea what would happen if I use my own LLVM instead of Julia's shared LLVM.

**Reply by alexreinking** [reply]
On Linux and macOS, where we have linker scripts that hide LLVM's symbols, it should be okay to link a static, custom LLVM to shared Halide. There's more work to be done on Windows, where ODR violations are still a legitimate concern.

**Reply by vchuravy** [reply]
From the Julia side we do strongly encourage users to stick with our patch-set otherwise you will eventually run into code-gen bugs.
LLVM12 hasn't been released yet, but we started the upgrading process (that takes time and will only be available on 1.7-dev)

From a deployment perspective you should likely build Halide in Yggdrasil and then have the wrapper use the Julia LLVM.

**Comment by johnnychen94** [comment]
Is it possible to pass the Halide Buffer as a Julia Array? I did some experiment by passing the pointer to internal `halide_buffer_t::host` data (`get_buffer_host`), and then use Julia's [`unsafe_wrap`](https://docs.julialang.org/en/v1/base/c/#Base.unsafe_wrap-Union{Tuple{N},%20Tuple{T},%20Tuple{Union{Type{Array},%20Type{Array{T,%20N}%20where%20N},%20Type{Array{T,%20N}}},%20Ptr{T},%20Tuple{Vararg{Int64,%20N}}}}%20where%20{T,%20N}) to reinterpret it:

```julia
julia> using Halide.CppBindings: _Buffer

julia> data = _Buffer.Buffer{Float32}((4, 4))
Halide.CppBindings._Buffer.BufferAllocated{Float32}(Ptr{Nothing} @0x0000000003005570)

julia> make_array(data) = unsafe_wrap(Array{Float32, 2}, Ptr{Float32}(_Buffer.get_buffer_host(data).cpp_object), (4, 4); own=true)

julia> A = make_array(data)
4×4 Matrix{Float32}:
 1.68832f25   1.66783f19  7.14788f22  1.13189f21
 1.06028f-38  1.76852f19  1.2867f22   7.00624f22
 1.10953f27   7.17676f31  1.15879f24  1.28391f22
 6.73472f22   1.81781f31  3.02213f32  1.15879f24
```

I'm really not sure if this is legal from Halide side because the ownership is unclear to me.

**Reply by johnnychen94** [reply]
An update on this:

I think it's a good start to share the Halide buffer object with Julia. (Haven't yet considered how to share a Julia array with Halide but it's on my todolist.)

This is the Julia side of the Buffer wrapper, I use `--->` to show the C++/Julia function mapping relationship

```julia
# Halide           ---> Julia
# dimensions  ---> dimensions
# dim               ---> dim # a simplified tuple with (min, extent, stride)
# fill                 ---> buffer_fill!
# [](const Buffer &b) -> uint8_t* {return b.get()->raw_buffer()->host;}   ---> get_buffer_host

Base.eltype(data::Buffer{T}) where T = T
Base.ndims(data::Buffer) = convert(Int, dimensions(data))
Base.size(data::Buffer) = ntuple(i->convert(Int, dim(data, i-1)[2]), ndims(data))
Base.fill!(data::Buffer, x) = buffer_fill!(data, convert(eltype(data), x))

function unsafe_wrap_from_buffer(data::Buffer{T}; own=false) where T<:Real
    AType = Array{T, Base.ndims(data)}
    data_ptr = Ptr{T}(get_buffer_host(data).cpp_object)
    return unsafe_wrap(AType, data_ptr, size(data); own=own)
end
```

Then, by filling values on the Halide side, and view it on the Julia side, the result says I'm interpreting
the data correctly.

```julia
julia> data = _Buffer.Buffer{Float32}((2, 5))
Halide.CppBindings._Buffer.BufferAllocated{Float32}(Ptr{Nothing} @0x000000000406e570)

julia> data_view = _Buffer.unsafe_wrap_from_buffer(data; own=false)
2×5 Matrix{Float32}:
 3.9f-44  NaN  7.27143f-38  0.0  5.70253f-38
 0.0      NaN  0.0          0.0  0.0

julia> fill!(data, 0) # it calls fill in the Halide side

julia> data_view # the Julia side array gets updated 🎉 
2×5 Matrix{Float32}:
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0
```

Whether `own` should be `true` or `false` isn't clear to me. According to the comment, it says Halide doesn't own this array, does this
mean that I can let Julia manage this part of memory by setting `own=true`?

https://github.com/halide/Halide/blob/cb78a6b093037a60abbc0085815e7e7bd1fe6bba/src/runtime/HalideRuntime.h#L1444-L1446

**Comment by abadams** [comment]
halide_buffer_t doesn't have any reference counting or anything, so it definitely doesn't own the data

Halide::Buffer and Halide::Runtime::Buffer own the data in the cases where they do the malloc, and don't own the data in the cases where they're given a pointer. So in your example above I think Halide Buffer owns the data, and Julia just gets a view of it (but Julia owns the Halide::Buffer object itself).

If you can hide the fact that Halide::Buffer even exists from Julia users and just make everything accept and return Julia arrays, that seems like it would be even better. In that case any Halide::Buffer objects constructed under the hood would be non-owning.

**Reply by alexreinking** [reply]
> If you can hide the fact that Halide::Buffer even exists from Julia users and just make everything accept and return Julia arrays, that seems like it would be even better. In that case any Halide::Buffer objects constructed under the hood would be non-owning.

Using native types is an explicit goal of this project. The Python bindings accept numpy arrays, for instance.

**Reply by johnnychen94** [reply]
Unlike Python, Julia itself has many array types(e.g., `Array`, `SubArray`, `OffsetArray`), so it would be more convenient to provide a Halide buffer view in Julia as a new array type, and directly applies native Julia functions on this buffer view.

I haven't figure out the way to pass arbitrary array type to Halide buffer via CxxWrap; I might end up with only support the Julia built-in `Array`.

**Comment by johnnychen94** [comment]
> To make a real working Halide demo (e.g., minimal JIT compiled example), there're more types that needed to be wrapped. My plan is to finish the concept validation with a working demo before April 1, and then come up with a detailed proposal and send it to the mentioned mail halide-gsoc2021-mentors@mit.edu

Here it comes (it's just a preview so the usage and information will be definitely get improved):

```julia
julia> f = Func("Blur");

julia> x, y = Var("x"), Var("y");

# Julia doesn't support assignment operator `=`
# a workaround is to make it a 0-dimensional array and override the `setindex!`[] operator
julia> f[x, y][] = x + y;

julia> result = realize(f, (4, 4))[0];

julia> result_view = unsafe_wrap_from_buffer(Int32, result)
4×4 Matrix{Int32}:
 0  1  2  3
 1  2  3  4
 2  3  4  5
 3  4  5  6
```

So at least the minimal JIT example works now and I haven't met the LLVM symbolic issue so far. I'll try the AOT example tomorrow and see if it works, but I want to investigate the BinaryBuilder system in the following few days and come up with the proposal.

---

One key difference that shows up here is that Julia uses 1-based indexing, so technically, the result in Julia should be

```julia
4×4 Matrix{Int64}:
 2  3  4  5
 3  4  5  6
 4  5  6  7
 5  6  7  8
```

Question: Should I automatically offset the indexes to suit the Julia 1-based assumption, or just note this difference somewhere in the documentation? I prefer to do auto offset because it would work more smoothly with other Julia array types.

**Comment by johnnychen94** [comment]
> So a separate repo is fine for rapid development, but it must be integrated into our repository in the end.

I still think we need one (maybe two) separate repos for the Julia bindings.

The first repo is for the Julia source codes (and potentially documentation), I still think it's necessary for the following two reasons:

- Julia's package registration system requires the [Registrator bot](https://github.com/JuliaRegistries/Registrator.jl) to make PRs to the official registry. This means that separate version control is needed here and thus a separate repo. Julia adopts a rapid rolling version in the whole ecosystem so releases per several months might not be a good strategy for `Halide.jl`.
- The documentation for `Halide.jl` will be generated by https://github.com/JuliaDocs/Documenter.jl, and the docs system seems not compatible with Halide's current ones. A separate repo(can be placed together with `Halide.jl`) is needed to host the documentation.

The (optional) second repo is to host the generated library built by BinaryBuilder. Currently, every binary dependency is built with JuliaLang's buildbot and released in a separate repo in https://github.com/JuliaBinaryWrappers with `*_jll` name. I understand that we want to keep all related codes and CI workflows under halide org, so I can try to figure out how to set up our own BinaryBuilder workflow, but I'm not sure if we want to host the `Halide_jll` repo in halide org or in JuliaBinaryWrappers org. If we want to host it in our halide org, then one more separate repo is needed. (Julia also version controls the jll artifacts)

---

For clarification, my final goal of this project is to let users do `pkg> add Halide` and get everything installed without any form of manual compilation. An example of this is `FFTW`:

```julia
(@v1.6) pkg> add FFTW
  Installing known registries into `tmp`
       Added registry `General` to `tmp/registries/General`
   Resolving package versions...
   Installed IntelOpenMP_jll ─ v2018.0.3+2
   Installed MKL_jll ───────── v2021.1.1+1
   Installed FFTW_jll ──────── v3.3.9+7
   Installed AbstractFFTs ──── v1.0.1
   Installed Reexport ──────── v1.0.0
   Installed JLLWrappers ───── v1.2.0
   Installed FFTW ──────────── v1.3.2
```

**Comment by rootjalex** [comment]
@johnnychen94 Is the email address linked on your github accurate? We should probably discuss writing the proposal

**Reply by johnnychen94** [reply]
Yes, and I've just sent an incomplete draft version of my proposal to the email address `halide-gsoc2021-mentors@mit.edu`. For whoever not on that mail list and interested in this project, here it is

[GSoC2021_Proposal_for_Halide.pdf](https://github.com/halide/Halide/files/6264655/GSoC2021_Proposal_for_Halide.pdf)

and a simple demo that utilizes both Halide and JuliaImages:

![image](https://user-images.githubusercontent.com/8684355/113712256-7d2c5a00-9718-11eb-90c7-8abb4490fe39.png)
